### PR TITLE
version: redesign output format for the dvc version command #3499

### LINKED ISSUE #3499: version: redesign output
Current `dvc version` output is ugly, we could definitely do better:

```
DVC version: 0.89.0+9e96dd
Python version: 3.7.0
Platform: Linux-4.15.0-88-generic-x86_64-with-debian-buster-sid
Binary: False
Package: None
Supported remotes: azure, gdrive, gs, hdfs, http, https, s3, ssh, oss
Cache: reflink - not supported, hardlink - supported, symlink - supported
Filesystem type (cache directory): ('ext4', '/dev/sda1')
Filesystem type (workspace): ('ext4', '/dev/sda1')
```

Related https://github.com/iterative/dvc/issues/3185

### PR DESCRIPTION
**Pull request by sahilbhosale63**
Fix #3499 redesign the output format for the `dvc version` command.

### CODE REVIEW

**Comment by skshetry** [comment]
@sahilbhosale63, thanks for the PR. CI is failing. Did you run pre-commit hooks as mentioned here: https://dvc.org/doc/user-guide/contributing/core.

Also, it'd be good to see the outputs for different DVC installations/repos. Thanks.

File: dvc/command/version.py [hunk 1]
```diff
@@ -129,7 +127,12 @@ def get_supported_remotes():
             if not tree_cls.get_missing_deps():
                 supported_remotes.append(tree_cls.scheme)
 
-        return ", ".join(supported_remotes)
+        if supported_remotes.__len__() == 9:
+            return "All remotes"
+        elif supported_remotes.__len__() == 1:
```
Comment (skshetry) [inline dvc/command/version.py:132]: In Python, it's not good to use dunder methods directly(`__len__`). You can just use `len()`.

File: dvc/command/version.py [hunk 2]
```diff
@@ -129,7 +127,12 @@ def get_supported_remotes():
             if not tree_cls.get_missing_deps():
                 supported_remotes.append(tree_cls.scheme)
 
-        return ", ".join(supported_remotes)
+        if supported_remotes.__len__() == 9:
```
Comment (skshetry) [inline dvc/command/version.py:130]: @sahilbhosale63, our support for remotes could change, it'd be great to find some way to get all the counts of the remotes that we support.

File: dvc/command/version.py [hunk 3]
```diff
@@ -115,10 +111,12 @@ def get_linktype_support_info(repo):
                 os.unlink(dst)
             except DvcException:
                 status = "not supported"
-            cache.append(f"{name} - {status}")
+
+            if status == "supported":
+                cache.append(f"{name}")
```
Comment (skshetry) [inline dvc/command/version.py:116]: No need of f-strings here.
```suggestion
                cache.append(name)
```

**Comment by sahilbhosale63** [comment]
Screenshot from 2020-06-26 21-54-06
@skshetry I want to commit the changes which I have made but when I run git commit some tests are failing. And also when I check if my changes are committed or not via git status then it shows "your branch is up to date with origin/master".

What's the problem? Can you please guide me here @skshetry?

File: dvc/command/version.py [hunk 4]
```diff
@@ -10,7 +10,7 @@
 from dvc.exceptions import DvcException, NotDvcRepoError
 from dvc.scm.base import SCMError
 from dvc.system import System
-from dvc.utils import is_binary, relpath
+from dvc.utils.__init__ import relpath
```
Comment (pmrowla) [inline dvc/command/version.py:13]: this should just be 

```py
from dvc.utils import relpath
```

File: dvc/command/version.py [hunk 5]
```diff
@@ -46,13 +45,11 @@ def run(self):
...
                 info.append(
-                    "Cache: {}".format(self.get_linktype_support_info(repo))
+                    "Cache types: " + self.get_linktype_support_info(repo)
                 )
                 if psutil:
                     fs_type = self.get_fs_type(repo.cache.local.cache_dir)
-                    info.append(
-                        f"Filesystem type (cache directory): {fs_type}"
-                    )
+                    info.append(f"Directory cache: {fs_type}")
```
Comment (pmrowla) [inline dvc/command/version.py:52]: `Directory cache` feels a bit off here. Seems like it should be `Cache directory:` or maybe just `Cache:`

File: dvc/command/version.py [hunk 6]
```diff
@@ -129,7 +128,12 @@ def get_supported_remotes():
             if not tree_cls.get_missing_deps():
                 supported_remotes.append(tree_cls.scheme)
 
-        return ", ".join(supported_remotes)
+        if len(supported_remotes) == len(TREES):
+            return "All remotes"
+        elif len(supported_remotes) == 1:
+            return supported_remotes
+        else:
+            return ", ".join(supported_remotes)
```
Comment (pmrowla) [inline dvc/command/version.py:136]: nitpicky, but you don't need `elif`/`else` here
```suggestion
        if len(supported_remotes) == 1:
            return supported_remotes
        return ", ".join(supported_remotes)
```

File: dvc/command/version.py (hunk 5 above)
Comment (skshetry) [reply dvc/command/version.py:52]: +1 for `Cache:`

File: dvc/command/version.py (hunk 6 above)
Comment (skshetry) [inline dvc/command/version.py:136]: @sahilbhosale63, the plan was to make `Supports` use both remotes and cache types. But, as we didnot go through that, I think it's okay to not go "All remotes" route. But, it's your call.

File: dvc/command/version.py (hunk 5 above)
Comment (skshetry) [reply dvc/command/version.py:52]: Also, we could change the order of this to be just above `Workspace`.

**Comment by skshetry** [comment]
@sahilbhosale63, @pmrowla, maybe we could do something like this:
```
Build Info: Python 3.7.7 on Darwin-19.5.0-x86_64-i386-64bit (pip)
Build Info: Python 3.7.7 on Darwin-19.5.0-x86_64-i386-64bit (binary)
```

Also, it would be good to rename `Build Info` to `Platform` which is what it is.

Or, could add it to the version: `DVC version: 1.0.0 (pip)`?

**Comment by pmrowla** [comment]
> Or, could add it to the version: `DVC version: 1.0.0 (pip)`?

This might work best just because the platform line is already going to be pretty long in a lot of cases (even without the installation method)

**Comment by sahilbhosale63** [comment]
@skshetry can you please review this patch?

File: dvc/command/version.py [hunk 7]
```diff
@@ -28,12 +28,11 @@ def run(self):
         from dvc.repo import Repo
 
         info = [
-            f"DVC version: {__version__}",
-            f"Python version: {platform.python_version()}",
-            f"Platform: {platform.platform()}",
-            f"Binary: {is_binary()}",
-            f"Package: {PKG}",
-            f"Supported remotes: {self.get_supported_remotes()}",
+            f"DVC version: {__version__} ({PKG})",
```
Comment (skshetry) [inline dvc/command/version.py:31]: I get:
```console
DVC version: 1.0.1+412f39 (None)
```
`PKG` can be None too. So, let's leave that out if it's None.

File: dvc/command/version.py [hunk 8]
```diff
@@ -46,13 +45,11 @@ def run(self):
             # `dvc config cache.shared group`.
             if os.path.exists(repo.cache.local.cache_dir):
                 info.append(
-                    "Cache: {}".format(self.get_linktype_support_info(repo))
+                    "Cache types: " + self.get_linktype_support_info(repo)
```
Comment (skshetry) [inline dvc/command/version.py:48]: Let's use `{}` formated string as above with just the addtion of `Cache types:`.

File: dvc/command/version.py [hunk 9]
```diff
@@ -46,13 +45,11 @@ def run(self):
...
                 )
                 if psutil:
                     fs_type = self.get_fs_type(repo.cache.local.cache_dir)
-                    info.append(
-                        f"Filesystem type (cache directory): {fs_type}"
-                    )
+                    info.append(f"Cache with {fs_type}")
```
Comment (skshetry) [inline dvc/command/version.py:52]: I agree with @pmrowla that it looks strange with mixed colon/no-colon output. Let's just use `Cache directory: ext4 on /dev/sda4` kind of output. Same for workspace too (Workspace directory?).

File: dvc/command/version.py [hunk 10]
```diff
@@ -139,7 +144,7 @@ def _get_dvc_repo_info(repo):
     if repo.root_dir != repo.scm.root_dir:
         return "dvc (subdir), git"
 
-    return "dvc, git"
+    return "dvc + git"
```
Comment (skshetry) [inline dvc/command/version.py:147]: We need to make this uniform, either we should make it all commas or all plus-es. Could just use comma for now.

File: dvc/command/version.py [hunk 11]
```diff
@@ -28,12 +28,11 @@ def run(self):
         from dvc.repo import Repo
 
         info = [
-            f"DVC version: {__version__}",
-            f"Python version: {platform.python_version()}",
-            f"Platform: {platform.platform()}",
-            f"Binary: {is_binary()}",
-            f"Package: {PKG}",
```
Comment (skshetry) [inline dvc/command/version.py:35]: @efiop, is `Binary` necessary? I assume we can deduct that through `PKG`, right?

**Comment by sahilbhosale63** [comment]
rsz_screenshot_from_2020-07-21_20-55-51

@skshetry @pmrowla Does this looks good??

File: dvc/command/version.py (hunk 11 above)
Comment (efiop) [reply dvc/command/version.py:35]: @skshetry Yeah, could totally live without `Binary`. PKG is enough.

File: dvc/command/version.py (hunk 10 above)
Comment (jorgeorpinel) [reply dvc/command/version.py:147]: Same for cache types which seems to use `/` right now.

**Comment by jorgeorpinel** [comment]
I just noticed this PR, thanks for the mention.

> Supports: All remotes

How about

Supported remotes: {List them always} — actually I think that's how it is now so maybe this doesn't need a change...

Also:
- `Cache types` -> `Supported cache types` ?
- `Workspace Directory` - lower case `directory`
- `Cache Directory` - same ^. And put after Workspace dir?
- `Repo: ` -> `Repo control: ` ?
---

Also keep in mind this may need updates to https://dvc.org/doc/command-reference/version or other docs after merge. Thanks

File: dvc/command/version.py (hunk 10 above)
Comment (shcheklein) [reply dvc/command/version.py:147]: +1 for commas (since / is a path separator)
+1 to check other occurrences

**Comment by shcheklein** [comment]
@sahilbhosale63 why do we need (`----------------`) separator?

How does the output looks like when no cache dir is available?
I don't see the install method ... is it None on your screenshot? should we use some human readable name?

**Comment by sahilbhosale63** [comment]
@jorgeorpinel  https://github.com/iterative/dvc/issues/3499

**Comment by sahilbhosale63** [comment]
> @sahilbhosale63 why do we need (`----------------`) separator?
> 
> How does the output looks like when no cache dir is available?
> I don't see the install method ... is it None on your screenshot? should we use some human readable name?

When there is no cache dir it gives a warning.

**Comment by jorgeorpinel** [comment]
Sounds like that issue number should be in the PR description? Linked so merging this closes it. See https://docs.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords

**Comment by sahilbhosale63** [comment]
> Sounds like that issue number should be in the PR description? Linked so merging this closes it. See https://docs.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords

Added #3499 issue number to the PR description with the **Fix** keyword.

**Comment by sahilbhosale63** [comment]
The Travis Build test failures are unrelated.

File: dvc/command/version.py [hunk 12]
```diff
@@ -27,13 +27,18 @@ class CmdVersion(CmdBaseNoRepo):
...
+
         info = [
-            f"DVC version: {__version__}",
-            f"Python version: {platform.python_version()}",
-            f"Platform: {platform.platform()}",
-            f"Binary: {is_binary()}",
-            f"Package: {PKG}",
-            f"Supported remotes: {self.get_supported_remotes()}",
+            f"DVC version: {__version__} {package}",
+            "--------------------------------- \n",
```
Comment (skshetry) [inline dvc/command/version.py:38]: No need of newline after the divider.

Comment (skshetry) [reply dvc/command/version.py:38]:
```suggestion
            "---------------------------------",
```

**Comment by efiop** [comment]
Thank you @sahilbhosale63 ! :pray:

**Comment by jorgeorpinel** [comment]
Created iterative/dvc.org/issues/1635 to update the docs.

**Pull request merged**